File: Modules/BuiltInAlgorithms/Sources/NewShuffleMergeSort.swift
import AlgorithmKit
import SortEngineKit

/// Ported from ArrayV's `NewShuffleMergeSort` — genuinely in-place merging via perfect-shuffle
/// permutations, based on the shuffle/unshuffle construction from Ellis & Markov's "In-place
/// associative computation" line of work (the class's own citations: the shuffle algorithm and its
/// time-complexity proof). `shuffle`/`unshuffle` interleave (or un-interleave) a range using
/// cycle-following over `i -> i*2 mod size`, glued together for arbitrary-length ranges via
/// `shuffleEasy`/`unshuffleEasy` blocks sized to the largest power of 3 that fits — `shuffleEasy`'s
/// `size` parameter is a cycle *modulus*, not an element count: it only ever touches the `size - 1`
/// real positions `start+1 ... start+size-1`, never `start+size` itself.
///
/// Reuses `RotateMergeSort`/`RotateLSDRadixSort`/`RotateMSDRadixSort`'s `rotate`/`multiSwap`
/// Gries-Mills block-rotation primitive verbatim for the "plain" rotations
/// (`rotate`/`rotateEqual`); `rotateShuffled*` are the same block-rotation shape but stepping by 2
/// (rotating already-shuffled interleaved pairs rather than plain elements).
///
/// Extends `IterativeTopDownMergeSort` in ArrayV purely to reuse its doubling-`subarrayCount`
/// driver loop with this file's own `merge` substituted in — this port keeps its own copy of that
/// driver rather than sharing a `Template` type with `IterativeTopDownMergeSort.swift`, the same
/// precedent the `i`/`b`/`q`/`m` radix family used (`StacklessBinaryQuickSort`/`RotateMSDRadixSort`):
/// only two real consumers exist, and this one genuinely overrides the interesting part rather than
/// purely delegating, so a shared enum wouldn't pull its weight. No aux buffer is needed here at
/// all (unlike the parent's own `merge`) since every move is either a swap or an in-place
/// permutation-cycle write.
///
/// Stability: not established. `mergeUp`'s `type` flag flips exactly when a shuffled run boundary
/// is crossed, which reads like deliberate stable-merge engineering, but confirming that by
/// inspection wasn't tractable in the time budgeted, and no valid empirical technique applies here
/// either (`rotate*` moves via `swap`, but `shuffleEasy`/`unshuffleEasy` move via `setValue` — a mix
/// the swap-tape-shadow technique this codebase's other stability tests rely on can't handle, the
/// same limitation `AmericanFlagSort`'s doc comment already explains). Left at the conservative
/// default rather than asserted either way.
public struct NewShuffleMergeSort: SortAlgorithm {
  public let id = AlgorithmID(rawValue: "newshufflemergesort")
  public let metadata = AlgorithmMetadata(
    displayName: "New Shuffle Merge Sort",
    category: .merge,
    sizeRange: 16...256,
    growthModel: OperationGrowthModel(
      anchorSize: 818, coefficients: [228622, 411.74, 0.115292],
      measuredSafeCeiling: nil),
    stable: false,
    timeComplexity: ComplexityBounds(
      best: "O(n log n)", average: "O(n log n)", worst: "O(n log n)"),
    spaceComplexity: "O(1)",
    iconName: "shuffle"
  )

  public init() {}

  public func record(into engine: inout RecordingEngine) {
    let n = engine.count
    guard n >= 2 else { return }

    func multiSwap(_ a: Int, _ b: Int, _ len: Int) {
      for i in 0..<len { engine.swap(a + i, b + i) }
    }

    func rotate(_ midIn: Int, _ aIn: Int, _ bIn: Int) {
      var mid = midIn
      var a = aIn
      var b = bIn
      while a > 0 && b > 0 {
        if a > b {
          multiSwap(mid - b, mid, b)
          mid -= b
          a -= b
        } else {
          multiSwap(mid - a, mid, a)
          mid += a
          b -= a
        }
      }
    }

    func shuffleEasy(_ start: Int, _ size: Int) {
      var i = 1
      while i < size {
        var val = engine.values[start + i - 1]
        var j = i * 2 % size
        while j != i {
          let nval = engine.values[start + j - 1]
          engine.setValue(start + j - 1, val)
          val = nval
          j = j * 2 % size
        }
        engine.setValue(start + i - 1, val)
        i *= 3
      }
    }

    func shuffle(_ startIn: Int, _ end: Int) {
      var start = startIn
      while end - start > 1 {
        let half = (end - start) / 2
        var l = 1
        while l * 3 - 1 <= 2 * half { l *= 3 }
        let m = (l - 1) / 2

        rotate(start + half, half - m, m)
        shuffleEasy(start, l)
        start += l - 1
      }
    }

    func rotateShuffledEqual(_ a: Int, _ b: Int, _ size: Int) {
      var i = 0
      while i < size {
        engine.swap(a + i, b + i)
        i += 2
      }
    }

    func rotateShuffled(_ midIn: Int, _ aIn: Int, _ bIn: Int) {
      var mid = midIn
      var a = aIn
      var b = bIn
      while a > 0 && b > 0 {
        if a > b {
          rotateShuffledEqual(mid - b, mid, b)
          mid -= b
          a -= b
        } else {
          rotateShuffledEqual(mid - a, mid, a)
          mid += a
          b -= a
        }
      }
    }

    func rotateShuffledOuter(_ midIn: Int, _ aIn: Int, _ bIn: Int) {
      var mid = midIn
      var a = aIn
      var b = bIn
      if a > b {
        rotateShuffledEqual(mid - b, mid + 1, b)
        mid -= b
        a -= b
        rotateShuffled(mid, a, b)
      } else {
        rotateShuffledEqual(mid - a, mid + 1, a)
        mid += a + 1
        b -= a
        rotateShuffled(mid, a, b)
      }
    }

    func unshuffleEasy(_ start: Int, _ size: Int) {
      var i = 1
      while i < size {
        var prev = i
        let val = engine.values[start + i - 1]
        var j = i * 2 % size
        while j != i {
          engine.setValue(start + prev - 1, engine.values[start + j - 1])
          prev = j
          j = j * 2 % size
        }
        engine.setValue(start + prev - 1, val)
        i *= 3
      }
    }

    func unshuffle(_ startIn: Int, _ end: Int) {
      var start = startIn
      while end - start > 1 {
        let half = (end - start) / 2
        var l = 1
        while l * 3 - 1 <= 2 * half { l *= 3 }
        let m = (l - 1) / 2

        rotateShuffledOuter(start + 2 * m, 2 * m, 2 * half - 2 * m)
        unshuffleEasy(start, l)
        start += l - 1
      }
    }

    // Tri-state compare, matching ArrayV's `Reads.compareIndices`: -1 (less), 0 (equal), 1
    // (greater). Records exactly one comparison op, inferring equality from the values already
    // read rather than issuing a second one.
    func compare3(_ i: Int, _ j: Int) -> Int {
      if engine.compare(i, j, by: (<)) { return -1 }
      return engine.values[i] == engine.values[j] ? 0 : 1
    }

    func mergeUp(_ start: Int, _ end: Int, _ typeIn: Bool) {
      var i = start
      var j = i + 1
      var type = typeIn
      while j < end {
        let cmp = compare3(i, j)
        if cmp == -1 || (!type && cmp == 0) {
          i += 1
          if i == j {
            j += 1
            type.toggle()
          }
        } else if end - j == 1 {
          rotate(j, j - i, 1)
          break
        } else {
          var r = 0
          if type {
            while j + 2 * r < end, compare3(j + 2 * r, i) != 1 { r += 1 }
          } else {
            while j + 2 * r < end, compare3(j + 2 * r, i) == -1 { r += 1 }
          }
          j -= 1
          unshuffle(j, j + 2 * r)
          rotate(j, j - i, r)
          i += r + 1
          j += 2 * r + 1
        }
      }
    }

    func merge(_ start: Int, _ mid: Int, _ end: Int) {
      if mid - start <= end - mid {
        shuffle(start, end)
        mergeUp(start, end, true)
      } else {
        shuffle(start + 1, end)
        mergeUp(start, end, false)
      }
    }

    func ceilPowerOfTwo(_ x: Int) -> Int {
      var x = x - 1
      var i = 16
      while i > 0 {
        x |= x >> i
        i >>= 1
      }
      return x + 1
    }

    var subarrayCount = ceilPowerOfTwo(n)
    while subarrayCount > 1 {
      var i = 0
      while i < subarrayCount {
        merge(n * i / subarrayCount, n * (i + 1) / subarrayCount, n * (i + 2) / subarrayCount)
        i += 2
      }
      subarrayCount >>= 1
    }
  }
}

File: App/Resources/AlgorithmDetails/newshufflemergesort/newshufflemergesort.java
import java.util.Arrays;

public class newshufflemergesort {
  private static void multiSwap(int[] arr, int i, int j, int length) {
    for (int k = 0; k < length; k++) {
      int t = arr[i + k];
      arr[i + k] = arr[j + k];
      arr[j + k] = t;
    }
  }

  private static void rotate(int[] arr, int mid, int leftLen, int rightLen) {
    while (leftLen > 0 && rightLen > 0) {
      if (leftLen > rightLen) {
        multiSwap(arr, mid - rightLen, mid, rightLen);
        mid -= rightLen;
        leftLen -= rightLen;
      } else {
        multiSwap(arr, mid - leftLen, mid, leftLen);
        mid += leftLen;
        rightLen -= leftLen;
      }
    }
  }

  // Perfect-shuffles a chunk of `size - 1` elements by following the cycles of i -> i*2 mod size.
  private static void shuffleBlock(int[] arr, int start, int size) {
    int i = 1;
    while (i < size) {
      int val = arr[start + i - 1];
      int j = (i * 2) % size;
      while (j != i) {
        int nextVal = arr[start + j - 1];
        arr[start + j - 1] = val;
        val = nextVal;
        j = (j * 2) % size;
      }
      arr[start + i - 1] = val;
      i *= 3;
    }
  }

  // A single riffle shuffle only closes into clean cycles at power-of-three sizes, so shuffling
  // happens in power-of-three chunks, rotating the next chunk's tail into place before each one.
  private static void shuffle(int[] arr, int start, int end) {
    while (end - start > 1) {
      int half = (end - start) / 2;
      int chunk = 1;
      while (chunk * 3 - 1 <= 2 * half) {
        chunk *= 3;
      }
      int tail = (chunk - 1) / 2;
      rotate(arr, start + half, half - tail, tail);
      shuffleBlock(arr, start, chunk);
      start += chunk - 1;
    }
  }

  private static void rotateShuffledEqual(int[] arr, int i, int j, int size) {
    for (int k = 0; k < size; k += 2) {
      int t = arr[i + k];
      arr[i + k] = arr[j + k];
      arr[j + k] = t;
    }
  }

  private static void rotateShuffled(int[] arr, int mid, int leftLen, int rightLen) {
    while (leftLen > 0 && rightLen > 0) {
      if (leftLen > rightLen) {
        rotateShuffledEqual(arr, mid - rightLen, mid, rightLen);
        mid -= rightLen;
        leftLen -= rightLen;
      } else {
        rotateShuffledEqual(arr, mid - leftLen, mid, leftLen);
        mid += leftLen;
        rightLen -= leftLen;
      }
    }
  }

  private static void rotateShuffledOuter(int[] arr, int mid, int leftLen, int rightLen) {
    if (leftLen > rightLen) {
      rotateShuffledEqual(arr, mid - rightLen, mid + 1, rightLen);
      mid -= rightLen;
      leftLen -= rightLen;
      rotateShuffled(arr, mid, leftLen, rightLen);
    } else {
      rotateShuffledEqual(arr, mid - leftLen, mid + 1, leftLen);
      mid += leftLen + 1;
      rightLen -= leftLen;
      rotateShuffled(arr, mid, leftLen, rightLen);
    }
  }

  // The inverse of shuffleBlock: walks the same cycles, writing each value one step backward.
  private static void unshuffleBlock(int[] arr, int start, int size) {
    int i = 1;
    while (i < size) {
      int prev = i;
      int val = arr[start + i - 1];
      int j = (i * 2) % size;
      while (j != i) {
        arr[start + prev - 1] = arr[start + j - 1];
        prev = j;
        j = (j * 2) % size;
      }
      arr[start + prev - 1] = val;
      i *= 3;
    }
  }

  private static void unshuffle(int[] arr, int start, int end) {
    while (end - start > 1) {
      int half = (end - start) / 2;
      int chunk = 1;
      while (chunk * 3 - 1 <= 2 * half) {
        chunk *= 3;
      }
      int tail = (chunk - 1) / 2;
      rotateShuffledOuter(arr, start + 2 * tail, 2 * tail, 2 * half - 2 * tail);
      unshuffleBlock(arr, start, chunk);
      start += chunk - 1;
    }
  }

  private static int compare3(int[] arr, int i, int j) {
    if (arr[i] < arr[j]) {
      return -1;
    }
    return arr[i] == arr[j] ? 0 : 1;
  }

  // Scans the shuffled (interleaved) range one adjacent pair at a time. A pair already in order
  // just advances the scan; a stretch of same-side elements gets un-shuffled back into two short
  // plain runs and rotated into its final position.
  private static void mergeUp(int[] arr, int start, int end, boolean fromLeft) {
    int i = start;
    int j = i + 1;
    while (j < end) {
      int cmp = compare3(arr, i, j);
      if (cmp == -1 || (!fromLeft && cmp == 0)) {
        i++;
        if (i == j) {
          j++;
          fromLeft = !fromLeft;
        }
      } else if (end - j == 1) {
        rotate(arr, j, j - i, 1);
        break;
      } else {
        int run = 0;
        if (fromLeft) {
          while (j + 2 * run < end && compare3(arr, j + 2 * run, i) != 1) {
            run++;
          }
        } else {
          while (j + 2 * run < end && compare3(arr, j + 2 * run, i) == -1) {
            run++;
          }
        }
        j--;
        unshuffle(arr, j, j + 2 * run);
        rotate(arr, j, j - i, run);
        i += run + 1;
        j += 2 * run + 1;
      }
    }
  }

  private static void merge(int[] arr, int start, int mid, int end) {
    if (mid - start <= end - mid) {
      shuffle(arr, start, end);
      mergeUp(arr, start, end, true);
    } else {
      shuffle(arr, start + 1, end);
      mergeUp(arr, start, end, false);
    }
  }

  private static int ceilPow2(int x) {
    x -= 1;
    for (int shift = 16; shift > 0; shift >>= 1) {
      x |= x >> shift;
    }
    return x + 1;
  }

  public static void sort(int[] arr) {
    int n = arr.length;
    if (n < 2) {
      return;
    }

    int subarrayCount = ceilPow2(n);
    while (subarrayCount > 1) {
      int i = 0;
      while (i < subarrayCount) {
        int lo = n * i / subarrayCount;
        int mid = n * (i + 1) / subarrayCount;
        int hi = n * (i + 2) / subarrayCount;
        merge(arr, lo, mid, hi);
        i += 2;
      }
      subarrayCount >>= 1;
    }
  }

  public static void main(String[] args) {
    int[] array = new int[] {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};
    sort(array);
    System.out.println(Arrays.toString(array));
  }
}
